src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -0,0 +1,113 @@
+
+        public DynamicControllerEndpointMatcherPolicy(DynamicControllerEndpointSelector selector)
+        {
+            if (selector == null)
+            {
+                throw new ArgumentNullException(nameof(selector));
+            }
+
+            _selector = selector;
+        }
+
+        public override int Order => int.MinValue + 100;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            if (!ContainsDynamicEndpoints(endpoints))
+            {
+                // Dynamic controller endpoints are always dynamic endpoints.
+                return false;
+            }
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var metadata = endpoints[i].Metadata.GetMetadata<DynamicControllerMetadata>();
+                if (metadata != null)
+                {
+                    // Found a dynamic controller endpoint
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public async Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // There's no real benefit here from trying to avoid the async state machine.
+            // We only execute on nodes that contain a dynamic policy, and thus always have
+            // to await something.
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                var endpoint = candidates[i].Endpoint;
+
+                var metadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
+                if (metadata == null)
+                {
+                    continue;
+                }
+
+                var matchedValues = candidates[i].Values;
+                var endpoints = await _selector.SelectEndpointsAsync(metadata.Values);
+                if (endpoints.Count == 0)
+                {
+                    // If there's no match this is a configuration error. We can't really check
+                    // during startup that the action you configured exists.
+                    throw new InvalidOperationException(
+                        "Cannot find the fallback endpoint specified by route values: " + 
+                        "{ " + string.Join(", ", metadata.Values.Select(kvp => $"{kvp.Key}: {kvp.Value}")) + " }.");
+                }
+
+                var replacement = endpoints[0];
```

Comment:
This seems sketchy if the above call doesn't throw and we can't guarantee that this always results in a single endpoint being selected. I might not understand all the details of how this works, but it feels like when you had two matching actions.

Reply:
Resolving this comment since I replied to this concern above.